core/src/main/java/com/linecorp/armeria/common/stream/StreamWriterOutputStream.java

```diff
@@ -0,0 +1,120 @@
+import com.linecorp.armeria.common.util.UnmodifiableFuture;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+
+final class StreamWriterOutputStream<T> extends OutputStream {
+
+    private final StreamWriter<T> writer;
+    private final Function<? super HttpData, ? extends T> httpDataConverter;
+    private ByteBuf buffer;
+    private final int maxBufferSize;
+    private boolean closed;
+    private CompletableFuture<Void> writeFuture = UnmodifiableFuture.completedFuture(null);
+
+    StreamWriterOutputStream(StreamWriter<T> writer,
+                             Function<? super HttpData, ? extends T> httpDataConverter,
+                             int maxBufferSize) {
+        requireNonNull(writer, "writer");
+        requireNonNull(httpDataConverter, "httpDataConverter");
+        checkArgument(maxBufferSize > 0, "maxBufferSize: %s (expected: > 0)", maxBufferSize);
+        this.writer = writer;
+        this.httpDataConverter = httpDataConverter;
+        this.maxBufferSize = maxBufferSize;
+        buffer = Unpooled.buffer(maxBufferSize, maxBufferSize);
+    }
+
+    @Override
+    public void write(int b) throws IOException {
+        ensureOpen();
+        buffer.writeByte(b);
+        maybeDrain();
+    }
+
+    @Override
+    public void write(byte[] bytes, int off, int len) throws IOException {
+        requireNonNull(bytes, "bytes");
+        checkArgument(off >= 0, "off: %s (expected: >= 0)", off);
+        checkArgument(len >= 0, "len: %s (expected: >= 0)", len);
+        ensureOpen();
+        while (len > 0) {
+            final int writableBytes = buffer.writableBytes();
+            buffer.writeBytes(bytes, off, Math.min(writableBytes, len));
+            off += writableBytes;
+            len -= writableBytes;
+            ensureOpen();
+            maybeDrain();
+        }
+    }
+
+    @Override
+    public void flush() throws IOException {
+        ensureOpen();
+        drain();
+    }
+
+    @Override
+    public void close() throws IOException {
+        if (closed) {
+            return;
+        }
+        try {
+            drain();
+        } finally {
+            writer.close();
+            closed = true;
+        }
+    }
+
+    private void ensureOpen() throws IOException {
+        if (closed || !writer.isOpen()) {
+            closed = true;
+            throw new IOException("Stream closed");
+        }
+    }
+
+    private void maybeDrain() throws IOException {
+        if (!buffer.isWritable()) {
+            drain();
+        }
+    }
+
+    private void drain() throws IOException {
+        if (!buffer.isReadable() || !writer.isOpen()) {
+            return;
+        }
+        writeFuture.join();
+        if (!writer.tryWrite(httpDataConverter.apply(HttpData.wrap(buffer)))) {
+            throw new IOException("Stream closed");
+        }
+        buffer = Unpooled.buffer(maxBufferSize, maxBufferSize);
+        writeFuture = writer.whenConsumed();
+    }
```

Comment:
> (https://github.com/line/armeria/pull/4186#discussion_r932851829) Alternatively, we may want to give users some freedom to not consider backpressure.
In this case, it may be a better idea to consider performing the StreamWriter#whenConsumed check whenever the buffer is drained.

@jrhee17 , I updated PR based on you suggestion!

`wrtier.whenConsumed().join()` on drain() is blocking call and **writer should be subscribed** to get completion on `whenConsumed().join()`, I also add `writer.subscirbe(..)` on tests.

Please check modification on test too! 🙇

Reply:
Sorry about the frequent design change 😅 

We discussed this API further, and we're thinking that`StreamWriterOutputStream`